### ISSUE TITLE
Tests/error/warning messages need to be written for Export to Excel feature

### ISSUE THREAD

**Issue by kdahlquist** [dondi/GRNsight#733]
The basic functionality of Export to Excel has been implemented in the UI.  This feature works as advertised when exporting from demo files 1 or 2, which were made from perfectly formed GRNmap-compatible workbooks.  However, for example, the export does not work for demo files 3 or 4, which are based on a data file for the 2015 Bulletin of Mathematical Biology paper, which were for an earlier version of the modeling software before it became GRNmap.

We need to implement a testing/error/warning system for export problems.  For GRNsight "error" messages are for things that are fatal errors, "warnings" are to alert the user that something unexpected might be happening.

If I had to guess what is happening with demos 3 and 4, it would be that not all the data expected for a full GRNmap workbook is there, so it's barfing.  

Bear in mind that if a user imports a SIF or GraphML file, then they would not have all the data for a full Excel export by definition.  But we should still allow them to export to Excel.  They would only have either a "network" worksheet or a "network_optimized_weights" worksheet.

So, bottom line, a "base" export would be of a "network" or "network_optimized_weights" worksheet.  Then, we should allow export of whatever other worksheet data is present, but not require that all the data be there, but take it on a worksheet by worksheet basis.  We could throw a warning to the user that says something like: "No data for <insert worksheet names> has been uploaded to GRNsight, so the exported Excel workbook will not contain those worksheets.

**Comment by afiller1** [comment]
I focused on this this week. There are a few things I've noticed and have questions on:
- Currently, attempting to Export to Excel for demos 3 and 4 cause the entire page to break. It transfers you over to localhost:5000 saying there's an error in the export_to_excel function.
- BUT. I found the section of code that seems to be causing this total breakage, it starts at line 86 in server/controllers/exporters/xlsx.js. When I comment this code out, everything seems to work fine.

I have a few questions, because I'm still somewhat unfamiliar with how exactly data is input into this tool and what kind of data is expected out of it given the type of input data.
- What's different about the data inputs for weighted vs unweighted graphs?
- What sheets should weighted vs unweighted Excel exports contain?
- Can we check the data output by the code I'm pushing to my branch, with the troublesome code commented out? Also, I want to ask Dondi whether this code is actually necessary in any case?

I have a few other code-specific questions to run by Dondi that should help me fix this problem altogether. I can still add warnings and error messages, but the fact that the whole site was crashing for me seemed like a more pertinent issue to resolve first. @kdahlquist @dondi

**Comment by kdahlquist** [comment]
@afiller1, you will need to become familiar with the GRNmap input/export format.  Here is a page from my biomath modeling class that talks about the GRNmap input format.  https://openwetware.org/wiki/BIOL388/S19:Week_6#Creating_the_GRNmap_Input_Workbook

There is also documentation on the GRNmap website here: http://kdahlquist.github.io/GRNmap/documentation/

Briefly, the output workbook needs to have everything that the input workbook has, plus some additional worksheets.

**Comment by dondi** [comment]
Upon examination of the line in question, @dondi suggested to really pinpoint the root cause of the problem. @afiller1 will come by during a future office hour to look at it.

**Comment by afiller1** [comment]
I didn't have time to stop by and work with Dondi (we were both at a conference this weekend). I pushed the code I have with the broken code commented out so we can try it out and look at it. I will go to Dondi's office hours today after class to resolve some of these problems.

**Comment by afiller1** [comment]
This week I worked on reading through the formatting rules for GRNmap files, I created new Excel files in test-files that can be used for error and warning testing, and I started on two new test files that will test 1. the imports of the network and network-weights in Excel files and 2. the expression data sheets. Those files are extremely rough right now, but looking through the code has helped me to understand how those tests work and how I will need to adjust them to fit Excel. I will come bother @dondi this week regarding this.

**Comment by kdahlquist** [comment]
A couple of notes:
* please make a list on the wiki of what you are testing so I can review and make suggestions
* keep the test files simple; they do not need to be as complicated as a real sample data file (like the demo files)
* one test per file

**Comment by afiller1** [comment]
Kinds of errors for expression data sheets:
- Wrong sheet naming format (STRAINNAME_log3_expression, for instance)
- Left column label: “id” ONLY
- Left column contains data of the format: 3 letters followed by a number (to represent the gene name)
	- Check for wrong naming format
	- Check for invalid name
- All other column headers are whole numbers. (Timepoints) (Only 15, 30, and/or 60)
- Throw error for non 15 or 30 or 60 number
  	- Throw error if not a number
- Same timepoints for all strains...But # replications may vary.
- Error if different timepoints (one only has 15, another only has 30, for instance)
 	- Make sure it accepts different # replications

Should I throw an error for missing data in a cell? Or should I treat it as a 0?

**Comment by Onariaginosa** [comment]
Something seems really off with the base Export to Excel functionality. At first I wanted to see what kinds of errors I could look for manually before writing a test to check for it. I ran a simple workbook that consists of only a network sheet, and the values (`test-files/graph-tests/different-sized-networks/10-genes-50-edges.xlsx` on the beta branch, `test-files/graph-tests/different-sized-networks/unused/10-genes-50-edges.xlsx` on the Onariaginosa branch) and I ran it through beta. 

I compared the two files and the network sheets are different. First the sheet names come out as 'workbook' and 'workbook_weights'. Additionally, the generated network data doesn't match the original network data. In many places the values are incorrect (there is a 0 where there should be a 1 and vice versa). Additionally, the 'optimization_parameters' sheet come out malformed, with only the column headers, and the row headers were 'data', 'errors', and 'warnings', with data having the value of '[object Object]'. 

I'm not sure how to structure the test to check whether the data is the same for a user exported file. Should I run the exported  excel file through `spreadsheet-controller` and compare the two network objects to see if there are any egregious differences between them (i.e network data not matching). That's the only way I can think of to test a user inputted workbook to an exported one, especially because people going from SIF or GraphML to Excel are not going to have all of the data that a normal workbook should have. 

Error Types:

Expression
- The ones mentioned above

Network
- Missing Source Genes
- Missing Target Genes
- Incorrect Genes (both Source and Target)
- Incorrect Matrix Data

Optimization Parameters
- Make sure that species information is there because we default to 'Saccharomyces cerevisiae'
- Taxon information

**Comment by kdahlquist** [comment]
For the "workbook" and "workbook" sheets problem that seems to be an artifact of the network/workbook substitutions that @igreen1 did.

This feature hasn't  been tested, so I'm not surprised that you are seeing fundamental issues with it.

**Comment by dondi** [comment]
In general, the track here is for @Onariaginosa to apply what has been done with errors and warnings but now for export. Possible approaches include using import tests to check whether an export correctly sends out data; another one suggested by @igreen1 is to look for a library that compares Excel files.

**Comment by Onariaginosa** [comment]
@Onariaginosa began writing export warning messages for different test cases. Right now they are all warning messages, but if the exported workbook returns a workbook that will throw errors if uploaded into GRNsight should they be warnings or errors? Additionally, if the workbook exports without semantic errors in it, but the data is just wrong (as mentioned in prior comments) should that be a warning that the data may be wrong, or should it be an error because the data is not correct?

**Comment by dondi** [comment]
Upon discussion of this issue, it appears the following cases may occur in terms of erroneous data:
* Bad data which the exporter does not detect—not much we can do here outside of gradually finding ways to detect them as we discover them
* (more pertinent to this issue) Bad data which the exporter detects—this breaks down into two more possibilities:
    * Fixable bad data—the exporter can go ahead and change the data in the export, perhaps issuing a warning/notification to the user that something was changed
    * Unfixable bad data—the exporter can detect a problem but there's no clear way to fix it automatically; these should be reported as errors, thus blocking the export and forcing the user to fix manually first

For the latter, @Onariaginosa will investigate further and if possible will provide some specific examples so that we can potentially walk through the data to pinpoint what to do.

**Comment by Onariaginosa** [comment]
Test 1: loaded Demo #1 file, exported demo
  Exported demo has no file errors
  Loaded Exported workbook:
- No errors/warnings

  Exported the previously exported graph:
    When attempted to load it throws 9 errors and refuses to load
    Malformed sheets:
  - optimization_parameters
  - production_rates 
  - degredation_rates
  - threshold_b
  

Test 2: loaded ./test-files/graph-tests/different-sized-networks/unused-files/10-genes-50-edges.xlsx
  File Description: Its just a network sheet
  Initial file warnings: 
  - Cell A1 in network is not 'cols regulators/rows targets'
    - Export should auto fix this warning 
  - No species information
    - Export should auto fix this warning as well
    - If species is changed export should reflect that
  - No expression sheet detected
      - If user selects a node coloring from the database, then export should reflect that
        - This occurs, but if you one dataset and then change it to something else, both appear in the export. Should this occur?
        - Additionally, files are exported without an expression sheet suffix, so when attempted to reload, the expression sheets are not recognized.
        
   Export file Issues:
  - network sheet has incorrect data
  -  network optimized weights exists (It was an unweighted export)
  - optimization parameters is malformed
    

Test 3: loaded ./test-files/additional-sheet-test-files/optimization-diagnostics-default.xlsx
  File Description: It is a properly formatted output workbook 
  No errors/warnings upon upload

  Sub-Test 1: Export Unweighted no changes
    File Issues: 
  - network_weights exists
  - network values are not 1s and 0s, and are incorrect.
  - optimization_parameters sheet is malformed
  - production_rates sheet is malformed
  - degradation_rates sheet is malformed
  - threshold_b rates sheet is malformed
    Attempts to reload exported file throws 10 errors

  Sub-Test2: Export Weighted no changes 
    File Issues:
  - network sheet values are not 1s and 0s
  - network_weights values are incorrect. (Bug) It appears that the values are being inserted row by row, when it should be column by column. The values themselves are correct, but in the wrong order. This is interesting because the expression sheets are all correct.
  - optimization_parameters sheet is malformed
  -  production_rates sheet is malformed
  - degradation_rates sheet is malformed
  - threshold_b rates sheet is malformed
  - output sheets like optimization_diagnostsics sheets disappear. Is this the correct behavior?

  Sub-Test3: Export Weighted change species
  - Because optimization_parameters sheet is consistently malformed, it is safe to say that we cannot check whether this change will occur

  Sub-Test4: Export Weighted change node coloring to use database expression sheet 
  - Changed the Top Dataset and Bottom Dataset to Barreto_2012_wt 
    - When exported, the same errors occurred as the Sub-Test2
    - In the export file, the Barreto_2012_wt sheet existed, but I don't believe that it contains an appropriate expression sheet suffix, so it may not appear when reloaded. I attempted to reload the export file to check, but it threw the same 10 errors as before.

Overview:
  - The `optimization_parameters`, `production_rates`, `degredation_rates`, `threshold_b` files are consistently malformed on export. 
  - The `network` sheet is created incorrectly, if there is a `network_optimized_weights` sheet. 
  - The values in both the `network` and the `network_optimized_weights` sheets are in the wrong places because they are inserted row by row instead of column by column
  - Output sheets are not in exported files even if they are imported (Is that correct?)
  - When exporting unweighted, make sure that there is only a `network` sheet, and that the values are restricted to only 0 or 1 in the matrix
  - When exporting data from the database, the new sheets dont have expression sheet suffixes, so upon reload they are not recognized

**Comment by dondi** [comment]
Issue #917 has been broken out of this issue so that it can be tackled separately.

**Comment by Onariaginosa** [comment]
I created tests that imports a workbook, exports it and imports the exported workbook to check whether there are no errors or warnings upon re-import for cases where we fixed a workbook with fixable warnings in the additional sheet files and cases where original file has no warnings/errors.

**Comment by dondi** [comment]
Work in progress, to appear after #920 is merged.

**Comment by Onariaginosa** [comment]
Created tests that imports a properly formatted workbook, exports it, and checks if the imported workbook is the same as the exported workbook.

**Comment by Onariaginosa** [comment]
See PR #925

**Comment by ahmad00m** [comment]
@ahmad00m confirms that the exported file is the same as the imported file. A file was exported then exported file was re-imported and then exported again and the exported file was the same as the re-imported files with no warning/errors.
